src/main.c

```diff
@@ -37,6 +45,17 @@ write_all(int fd, const char *buf, int n)
   return 0;
 }
 
+/*
+ * Set a socket to be nonblocking.
+ */
+int
+nonblock(int sockfd)
+{
+  int flags;
+  flags = fcntl(sockfd, F_GETFL, 0);
```

Comment:
Can this call ever fail? You should check if it returns `-1` before possibly mangling the flags on `sockfd`.